### PR TITLE
Abstracts out deployment execution to allow for extensibility

### PR DESCRIPTION
**Pull request by paynecrl97**
I looked into implementing Media export steps for deployment and realised that it wasn't possible to extend how the deployment packages are imported.

This is a small refactoring to add an `IDeploymentTargetHandler` interface that any module can implement to add functionality to the import logic. There's only one implementation at the minute- recipes, but this change now opens up the door for more (e.g. media import).

### CODE REVIEW

**Comment by sebastienros** [comment]
I understand that I added a way to add files to the archive during Export, but no way to do anything else with this archive during import, and this is what you are fixing here, correct? So we can already have deployment step that will export media files (the same way we add a recipe.json).

Can we also use this opportunity to fix other issues I see while reading the code (I am the culprit) ?
- Store a has of the API Key. Leaking its value is as bad as leaking a password otherwise.
- Don't allocate the zip file in memory (base64 conversion), but instead use the `FormFile` type from MVC to stream the content of the request. Also send it correctly from the client (I have some sample code in another app that will help).
- Allow for multiple recipe.json files (based on a folder? a filename convention like xyz.recipe.json?)
- Clean the temp folder when the import is done.
- Add a way to upload an export file, and maybe also a recipe file directly.

I am willing to do most of it based on your availability.

**Comment by paynecrl97** [comment]
Yeah, this PR provides a way for modules to contribute to what happens when a deployment is imported. Each registered handler gets given an instance of the imported folder, so they can do whatever they need to do with the files in it.

I would add to those changes:

- Move the orchestration of the import to the `Deployment` module (it's currently in a controller in `Deployment.Remote`).

I have a small amount of time tonight, but apart from that I won't have much time to work on this for the next few weeks. I'll see what I can get done today and then we can work out how to split the work.

**Comment by paynecrl97** [comment]
OK, I didn't have much time, so I did the two easiest changes-

- Move orchestration from the remote controller into a deployment service
- Provide UI to upload a package

Some of the other changes listed above may alter the interfaces slightly, but I think this is a good place to merge before we start the other work (it unblocks development of the media file import/export, and also adds a useful feature in the upload UI).

I'm away for the next few days so won't be able to pick up any changes etc...

File: src/OrchardCore.Modules/OrchardCore.Deployment/AdminMenu.cs [hunk 1]
```diff
@@ -28,6 +28,11 @@ public void BuildNavigation(string name, NavigationBuilder builder)
                             .Permission(Permissions.Export)
                             .LocalNav()
                         )
+                        .Add(T["Package Import"], "5", deployment => deployment
```
[the remark this answers is not in the document]
Comment (sebastienros) [reply src/OrchardCore.Modules/OrchardCore.Deployment/AdminMenu.cs]: Talked with our security guru, he agrees with my suggestion, but these permissions should not be granted by default to any role, they should be set by the site owner only, as after all only the site owner can change the site owner.

But in the case of running an imported recipe, the steps are executed under the account of the executor, so maybe we just need to ensure that every Step is checking for the correct permissions. And also ensure that the Import action impersonates the request with a specific role, or user (or site owner?) so we can limit what imported recipes can do.

**Comment by paynecrl97** [comment]
Yep, that makes sense. 

What do we do for assigning permissions in the roles UI? I know in O1 that a user who could edit roles was able to assign the site owner role to a role that they themselves had, which would allow them free reign on the site. Maybe we change this so that they can only assign permsissions that they already have etc?

**Pull request merged**